sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,443 @@
+        if buffer.len() < required_len {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        Ok(num_entries)
+    }
+
+    /// Fetches the SlotHashes sysvar data directly via syscall into a provided buffer
+    /// without validation.
+    ///
+    /// This method is for programs that cannot include the sysvar account
+    /// but still need access to the slot hashes data.
+    ///
+    /// # Arguments
+    /// * `buffer` - A mutable slice to store the sysvar data. The buffer length
+    ///   determines how much data is fetched. Use 20,488 bytes for full data
+    ///   on mainnet.
+    /// * `offset` - Byte offset within the sysvar data to start fetching from.
+    ///   Note: SlotHashes data starts with an 8-byte length prefix followed by entries.
+    ///   Must be 0 (start of data) or 8 + N*40 (start of entry N) for valid alignment,
+    ///   but this is not checked.
+    ///
+    /// # Returns
+    /// Nothing - the caller constructs the SlotHashes view afterwards.
+    ///
+    /// For most use cases, prefer `from_account_info()` which provides zero-copy access.
+    #[inline(always)]
+    pub fn fetch_into_unchecked(buffer: &mut [u8], offset: usize) -> Result<(), ProgramError> {
+        // Fetch sysvar data into caller-provided buffer.
+        unsafe {
+            crate::sysvars::get_sysvar_unchecked(
+                buffer.as_mut_ptr(),
+                &SLOTHASHES_ID,
+                offset,
+                buffer.len(),
+            )
+        }?;
+
+        Ok(())
+    }
+
+    /// Returns the number of `SlotHashEntry` items accessible.
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Returns `true` if there are no entries.
+    #[inline(always)]
+    pub fn is_empty(&self) -> bool {
+        self.len == 0
+    }
+
+    /// Returns the entire slice of entries. Call once and reuse the slice if you
+    /// need many look-ups.
+    #[inline(always)]
+    pub fn entries(&self) -> &[SlotHashEntry] {
+        self.as_entries_slice()
+    }
+
+    /// Gets a reference to the entry at `index` or `None` if out of bounds.
+    #[inline(always)]
+    pub fn get_entry(&self, index: usize) -> Option<&SlotHashEntry> {
+        self.as_entries_slice().get(index)
+    }
+
+    /// Finds the hash for a specific slot using binary search.
+    ///
+    /// Returns the hash if the slot is found, or `None` if not found.
+    /// Assumes entries are sorted by slot in descending order.
+    /// If calling repeatedly, prefer getting `entries()` in caller
+    /// to avoid repeated slice construction.
+    #[inline(always)]
+    pub fn get_hash(&self, target_slot: Slot) -> Option<&[u8; HASH_BYTES]> {
+        let entries = self.as_entries_slice();
+        entries
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+            .map(|index| &entries[index].hash)
+    }
```

Comment:
*minor*

Some duplication can be removed, if we use `position()` here:

```suggestion
    pub fn get_hash(&self, target_slot: Slot) -> Option<&[u8; HASH_BYTES]> {
        let entries = self.as_entries_slice();
        self.position().map(|index| &entries[index].hash)
    }
```

Reply:
Thanks, appreciate it. This is back into draft as I'm benchmarking a lot of tiny differences, so it will see a bunch of (probably) tiny adjustments over the next few.